### PR TITLE
Implement parallel doublewrite

### PR DESCRIPTION
**Pull request by laurynas-biveinis**
Blueprint:
https://blueprints.launchpad.net/percona-server/+spec/parallel-doublewrite

New data structure for a single doublewrite partition: struct
parallel_dlbwr_shard_t, and a new structure for the whole doublewrite:
struct parallel_dblwr_t.

Remove batch doublewrite fields from struct buf_dblwr_t: first_free,
b_reserved, b_event, and batch_running.

New methods to support parallel doublewrite:
buf_parallel_dblwr_make_path, buf_parallel_dblwr_close,
buf_parallel_dblwr_delete, buf_parallel_dblwr_finish_recovery,
buf_parallel_dblwr_file_create, buf_parallel_dblwr_create,
buf_parallel_dblwr_destroy, buf_parallel_dblwr_partition,
buf_parallel_dblwr_shard_num.  New global variable
parallel_dblwr_buf.

Make buf_dblwr_init_or_load_pages to read any pages from the parallel
doublewrite file.

Delete the old one if existing and re-create the doublewrite buffer at
the end of buf_dblwr_process.

Make buf_dblwr_update to decrement atomically the remaining to process
page count for the batch flushing modes, and signal the doublewrite
partition event on the last one.

Make buf_dblwr_flush_buffered_writes take a doublewrite partition
index argument, remove the logic to sync with an already running
batch.

Make buf_dblwr_add_to_batch take a flush type argument in order to
find the correct doublewrite shard for the page being added. Remove
all logic to sync with an already running batch.

In buf_dblwr_update and buf_dblwr_flush_single_page, allow single page
flushes to take any slot in the legacy doublewrite buffer.  Simplify
the latter function.

Remove any synchronous flushing code so that no thread except from the
cleaner/LRU manager threads can do any kind of flushing by
itself. Instead, any other thread has to wait for the flusher threads
to do the required flushing. This affects buf_flush_sync_all_pools,
and log_preflush_pool_modified_pages.

Do not make log_checkpoint_at flush at the very end of server shutdown
when neither the page cleaner is running nor there can be dirty pages
in the buffer pool.

Destroy (and delete its file) the parallel doublewrite at the end of
logs_empty_and_mark_files_at_shutdown.

Delete the existing doublewrite buffer, if any, in
recv_recovery_from_checkpoint_start, if found that the instance does
not need crash recovery.

Create the doublewrite buffer in
innobase_start_or_create_for_mysql. Add an assert that any pages from
the old buffer have been used or it was impossible to use them due to
some condition (an error, high innodb_force_recovery value, etc).

Introduce new server variable innodb_parallel_doublewrite_path.  Allow
UNIV_DEBUG || UNIV_PERF_DEBUG server variable
innodb_doublewrite_batch_size to go to MAX_DOUBLEWRITE_BATCH_SIZE
(currently defined as 200) maximum instead of the current 127.

Register the parallel doublewrite file with Performance Schema.

Fix bug 1549301 / http://bugs.mysql.com/bug.php?id=80496
(buf_dblwr_init_or_load_pages now returns an error code, but caller
not updated) as the function in question now can return many more
errors related to the parallel doublewrite file I/O.

Move suite/innodb/t/doublewrite.test to
include/doublewrite.inc. Create new innodb.doublewrite testcase that
simply includes that file and so has the same effect. Also create
innodb.doublewrite_multiple_pools that includes the same, but has two
buffer pool instances so that the tested data pages fall into
different doublewrite shards.

Create new testcase innodb.percona_doublewrite that tests the
doublewrite file handling, path setting, and error detection.

Create new testcase innodb.percona_doublewrite_single_page_flush that
tests the single page flush code path of the legacy doublewrite.

Update perfschema.show_sanity and create
sys_vars.innodb_percona_doublewrite_path testcase for the new
variable.

Make include/start_mysqld.inc and restart_mysqld.inc accept a one-shot
option $restart_hide_args, that would not print the actual server args
after the "# restart" string. This is done in order to test server
restart where the passed args should not be recorded in the output,
such as temp file paths.

Remove xb_doublewrite file in selected testcases manually as needed.

http://jenkins.percona.com/job/mysql-5.7-param/97/
http://jenkins.percona.com/job/mysql-5.7-param/98/#showFailuresLink for a single testcase (perfschema.show_sanity)

### CODE REVIEW

File: storage/innobase/buf/buf0dblwr.cc [hunk 1]
```diff
@@ -338,6 +336,92 @@ buf_dblwr_create(void)
...
+dberr_t
+buf_parallel_dblwr_make_path(void)
+{
+	if (parallel_dblwr_buf.path)
+		return(DB_SUCCESS);
+
+	char path[FN_REFLEN];
+
+	if (srv_parallel_doublewrite_path
+	    && srv_parallel_doublewrite_path[0] == OS_PATH_SEPARATOR) {
```
Comment (gl-sergei) [inline storage/innobase/buf/buf0dblwr.cc]: is_absolute_path should work here

File: storage/innobase/buf/buf0dblwr.cc [hunk 2]
```diff
@@ -338,6 +336,92 @@ buf_dblwr_create(void)
...
+	if (ret && exists) {
+		if (my_realpath(path, path, MY_WME) != 0)
+			return(DB_ERROR);
+		if (type != OS_FILE_TYPE_FILE) {
+			ib::error() << "Parallel doublewrite path "
+				    << path << " must point to a regular "
+				"file";
+			return(DB_WRONG_FILE_NAME);
+		}
+	} else if (srv_parallel_doublewrite_path[0] != OS_PATH_SEPARATOR) {
```
Comment (gl-sergei) [inline storage/innobase/buf/buf0dblwr.cc]: same above, is_absolute_path better to use

File: storage/innobase/buf/buf0dblwr.cc [hunk 3]
```diff
@@ -338,6 +336,92 @@ buf_dblwr_create(void)
...
+		resolve only srv_data_home and append
+		srv_parallel_doublewrite_path to it. */
+		char	srv_data_home_full[FN_REFLEN];
+
+		if (my_realpath(srv_data_home_full, srv_data_home, MY_WME)
+		    != 0) {
+			return(DB_ERROR);
+		}
+
+		if (srv_data_home_full[strlen(srv_data_home_full) - 1]
```
Comment (gl-sergei) [inline storage/innobase/buf/buf0dblwr.cc]: shouldn't my_realpath always remove extra path separator?

**Comment by gl-sergei** [comment]
buf0dblwr.ic lacks file header

**Comment by gl-sergei** [comment]
Full review yet to be done, above is just what I have found after quick look.

File: storage/innobase/buf/buf0dblwr.cc (hunk 3 above)
Comment (laurynas-biveinis) [inline storage/innobase/buf/buf0dblwr.cc]: I don't know: http://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html and http://man7.org/linux/man-pages/man3/realpath.3.html are not clear to me with this respect.

**Comment by laurynas-biveinis** [comment]
Not only the .ic file lacks the header file, it looks like I forgot to update copyrights for everything touched in 2016 (and 2015!)

File: storage/innobase/handler/ha_innodb.cc [hunk 4]
```diff
@@ -19727,7 +19736,7 @@ static MYSQL_SYSVAR_ULONG(page_hash_locks, srv_n_page_hash_locks,
 static MYSQL_SYSVAR_ULONG(doublewrite_batch_size, srv_doublewrite_batch_size,
   PLUGIN_VAR_OPCMDARG | PLUGIN_VAR_READONLY,
   "Number of pages reserved in doublewrite buffer for batch flushing",
-  NULL, NULL, 120, 1, 127, 0);
+  NULL, NULL, 120, 1, MAX_DOUBLEWRITE_BATCH_SIZE, 0);
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.cc:20108]: Cannot raise this above 127 without decoupling the legacy doublewrite buffer area size

File: mysql-test/suite/perfschema/t/show_sanity.test [hunk 5]
```diff
@@ -501,6 +501,7 @@ insert into test.sanity values
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_PAGE_CLEANER_DISABLED_DEBUG"),
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_PAGE_HASH_LOCKS"),
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_PAGE_SIZE"),
+  ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "INNODB_PARALLEL_DOUBLEWRITE_PATH"),
```
Comment (laurynas-biveinis) [inline mysql-test/suite/perfschema/t/show_sanity.test:504]: Incorrect since this is not "a known 5.6 system variable"

Comment (laurynas-biveinis) [inline mysql-test/suite/perfschema/t/show_sanity.test:504]: Hm, not sure, innodb_log_checksums is also there but it's definitely not a 5.6 variable

**Comment by laurynas-biveinis** [comment]
All comments so far addressed
http://jenkins.percona.com/job/mysql-5.7-param/101/#showFailuresLink

**Comment by gl-sergei** [comment]
Looks good to me

File: storage/innobase/buf/buf0dblwr.cc [hunk 6]
```diff
@@ -514,6 +599,104 @@ buf_dblwr_init_or_load_pages(
 		page += univ_page_size.physical();
 	}
 
+	err = buf_parallel_dblwr_make_path();
+	if (err != DB_SUCCESS)
+		return(err);
+
+	ut_ad(parallel_dblwr_buf.file == OS_FILE_CLOSED);
+	bool success;
+	parallel_dblwr_buf.file
```
Comment (laurynas-biveinis) [inline storage/innobase/buf/buf0dblwr.cc:624]: The file is read once here, sequentially, and closed. Probably os_file_setnocache makes sense too

File: storage/innobase/buf/buf0dblwr.cc [hunk 7]
```diff
@@ -338,6 +337,92 @@ buf_dblwr_create(void)
...
+	if (parallel_dblwr_buf.path)
+		return(DB_SUCCESS);
+
+	char path[FN_REFLEN];
+
+	if (srv_parallel_doublewrite_path
+	    && is_absolute_path(srv_parallel_doublewrite_path)) {
+
+		strncpy(path, srv_parallel_doublewrite_path, sizeof(path));
+	} else if (srv_data_home[strlen(srv_data_home) - 1]
```
Comment (laurynas-biveinis) [inline storage/innobase/buf/buf0dblwr.cc]: srv_data_home might be "" legitimately, see http://bugs.mysql.com/bug.php?id=78831

File: storage/innobase/buf/buf0dblwr.cc [hunk 8]
```diff
@@ -338,6 +337,92 @@ buf_dblwr_create(void)
...
+				"file";
+			return(DB_WRONG_FILE_NAME);
+		}
+	} else if (!is_absolute_path(srv_parallel_doublewrite_path)) {
+		/* If it does not exist, and is not an absolute path, then
+		resolve only srv_data_home and append
+		srv_parallel_doublewrite_path to it. */
+		char	srv_data_home_full[FN_REFLEN];
+
+		if (my_realpath(srv_data_home_full, srv_data_home, MY_WME)
```
Comment (laurynas-biveinis) [inline storage/innobase/buf/buf0dblwr.cc]: Likewise

**Comment by laurynas-biveinis** [comment]
```
http://jenkins.percona.com/job/mysql-5.7-param/112/
```

**Pull request merged**